pallets/unified-accounts/src/lib.rs

```diff
@@ -0,0 +1,404 @@
+            signature: EvmSignature,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            // make sure no prior mapping exists
+            ensure!(
+                !EvmToNative::<T>::contains_key(&who),
+                Error::<T>::AlreadyMapped
+            );
+            ensure!(
+                !NativeToEvm::<T>::contains_key(evm_address),
+                Error::<T>::AlreadyMapped
+            );
+
+            // recover evm address from signature
+            let address = Self::verify_signature(&who, &signature)
+                .ok_or(Error::<T>::UnexpectedSignatureFormat)?;
+
+            ensure!(evm_address == address, Error::<T>::InvalidSignature);
+
+            // Check if the default account id already exists for this evm address
+            let default_account_id = T::DefaultAddressMapping::into_account_id(evm_address.clone());
+            if frame_system::Pallet::<T>::account_exists(&default_account_id) {
+                // Transfer all the free native balance from old account id to the newly
+                // since this `default_account_id` will no longer be connected to evm address
+                // and users cannot access it.
+                // For the reset of the assets types (like XC20, etc) that should be handled by UI.
+                T::Currency::transfer(
+                    &default_account_id,
+                    &who,
+                    T::Currency::reducible_balance(&default_account_id, Expendable, Polite),
+                    Expendable,
+                )?;
+            }
+
+            // create double mappings for the pair
+            Self::add_mappings(who, evm_address);
+            Ok(())
+        }
+
+        /// Claim default evm address for given account id
+        /// Ensure no prior mapping exists for the account
+        ///
+        /// WARNINGS: Once connected user cannot change their mapping EVER.
+        #[pallet::call_index(1)]
+        #[pallet::weight(T::WeightInfo::claim_default_evm_address())]
+        pub fn claim_default_evm_address(origin: OriginFor<T>) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            // claim default evm address
+            let _ = Self::do_claim_default_evm_address(who)?;
+            Ok(())
+        }
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    /// Add the given pair to create double mappings
+    fn add_mappings(account_id: T::AccountId, evm_address: EvmAddress) {
+        NativeToEvm::<T>::insert(&evm_address, &account_id);
+        EvmToNative::<T>::insert(&account_id, &evm_address);
+
+        Self::deposit_event(Event::AccountClaimed {
+            account_id,
+            evm_address,
+        });
+    }
+
+    /// Claim the default evm address
+    fn do_claim_default_evm_address(account_id: T::AccountId) -> Result<EvmAddress, DispatchError> {
+        ensure!(
+            !EvmToNative::<T>::contains_key(&account_id),
+            Error::<T>::AlreadyMapped
+        );
```

Comment:
Suggestion for safer code:
* move the `ensure!` checks into `add_mappings` and make it return `Result<...>`.

Maybe I just find it strange that `add_mappings` exists though.
It's very simple & could just be part of the extrinsic directly.

If you don't want to modify this, please at least add a comment to the `add_mappings` that it doesn't perform any checks.

Reply:
I'll inline the `add_mappings()` method logic, it's only used 2 times anyway.